include/swift/AST/ClangModuleLoader.h

```diff
@@ -100,6 +101,15 @@ class ClangModuleLoader : public ModuleLoader {
   lookupRelatedEntity(StringRef clangName, ClangTypeKind kind,
                       StringRef relatedEntityKind,
                       llvm::function_ref<void(TypeDecl *)> receiver) = 0;
+
+  /// Try to parse the string as a Clang function type.
+  ///
+  /// Returns null if there was a parsing failure.
+  virtual const clang::Type *parseClangFunctionType(StringRef type) const = 0;
```

Comment:
Does this actually enforce that it's a function type?  Seems reasonable for the general routine to just parse any old type, and then the caller can do some simple checks like that it's a function type with the right number of arguments.

Reply:
> Does this actually enforce that it's a function type? Seems reasonable for the general routine to just parse any old type, and then the caller can do some simple checks like that it's a function type with the right number of arguments

The problem is the dependency thing. We don't want to pull in the Clang AST in our AST stuff (right?), so the check is in `ClangModuleLoader.cpp` instead. Otherwise, I wouldn't have jumped through so many hoops.

Reply:
Well, for now it's not doing any harm and we don't need to generalize.  It's true that we should endeavor to keep the Clang AST out of as much of the compiler as possible, and that most of the places that do need Clang should be talking through an abstract interface.

Reply:
Ok... so, I guess I don't understand the request here. Should I get rid of these two functions from the ClangModuleLoader? And have `Types.cpp`/`ASTContext.cpp` (which contain the printing functions) directly import the headers to implement the print methods? Or even have a getter for the field and import the Clang AST directly in the ASTPrinter?

Seems a little bit opposite in spirit to everything else I've done so far 😐.

Reply:
I think the whole spirit of generalization is already kinda' not there, since we just have one concrete class. Like, we have `ClangModuleLoader` giving us a `Sema`, nothing else apart from `ClangImporter` can really match that interface, can it?

Reply:
I was requesting that we just parse any old type, since we might want this functionality elsewhere in the future, and then the caller could validate whatever it wants to validate.  But then I relented, because you're not wrong that we may want to keep dependencies on the Clang AST out of Sema as much as possible.

I was not thinking about there being multiple implementations of `ClangModuleLoader`; I agree that that's somewhat silly, and I think we use `virtual` methods purely to avoid a link-time dependency on the importer.

I think the potential generalization direction in the future that we should keep in mind is being able to import from a broader set of (not-necessarily-C-based) languages.

Reply:
Understood. Marking this as resolved as I think I addressed the main question.